src/main/python/ttconv/scc/context.py

```diff
@@ -0,0 +1,479 @@
+# -*- coding: UTF-8 -*-
+
+# Copyright (c) 2020, Sandflow Consulting LLC
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""SCC context"""
+
+from __future__ import annotations
+
+import copy
+from typing import Optional, Type, Tuple
+
+from ttconv.model import Div
+from ttconv.scc.caption_paragraph import SccCaptionParagraph
+from ttconv.scc.caption_style import SccCaptionStyle
+from ttconv.scc.codes.attribute_codes import SccAttributeCode
+from ttconv.scc.codes.control_codes import SccControlCode
+from ttconv.scc.codes.mid_row_codes import SccMidRowCode
+from ttconv.scc.codes.preambles_address_codes import SccPreambleAddressCode
+from ttconv.scc.config import SccReaderConfiguration, TextAlignment
+from ttconv.scc.word import SccWord
+from ttconv.style_properties import StyleProperties
+from ttconv.time_code import SmpteTimeCode
+
+ROLL_UP_BASE_ROW = 15
+
+
+class SccContext:
+  """SCC context for reader"""
+
+  def __init__(self, safe_area_x_offset: int, safe_area_y_offset: int, config: Optional[SccReaderConfiguration] = None):
+    # Caption paragraphs container
+    self.div: Optional[Div] = None
+
+    # Caption paragraphs counter
+    self.count: int = 0
+
+    # Screen safe area offsets
+    self.safe_area_x_offset = safe_area_x_offset
+    self.safe_area_y_offset = safe_area_y_offset
+
+    # Previously read SCC word value
+    self.previous_word: Optional[SccWord] = None
+    self.previous_word_type: Optional[Type] = None
+
+    # Buffered caption being built
+    self.buffered_caption = None
+    # Captions being displayed
+    self.active_caption: Optional[SccCaptionParagraph] = None
+    # Caption style (Pop-on, Roll-up, Paint-on) currently processed
+    self.current_style = SccCaptionStyle.Unknown
```

Comment:
Shouldn't we have a "default" style?

Reply:
What would be this default style? It seems a bit tricky to choose...

Reply:
I skimmed the 608 spec and couldn't find an obvious answer.

I would choose paint-on for now.